### PR TITLE
AIRCoP dropout

### PR DESCRIPTION
**Pull request by luke-jiang**
Implement AIRCoP dropout

- [x] change dropout forward node from TENSOR to TENSORS
- [ ] factor out redundant code in dropout fwd and bwd passes
- [x] test dropout code on server
- [x] change `overwriteCheckFiles` to false before merge

### CODE REVIEW

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala [hunk 1]
```diff
@@ -72,6 +72,25 @@ trait FixedSizeDistributedTensorConvTypeLess extends FixedSizeDistributedTensorM
     TensorType(output_shape, input.et, anno)
   }
 
+  def DropoutForward(input: TENSOR, params: DropoutParam, anno: Anno, __pos: SourceContext): TENSOR = {
+    val output_tt = input.resultType
+    System.out.println(output_tt)
```
Comment (luke-jiang) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala:77]: printed type: `TensorType(List(Size(Dim(4),2), Size(Dim(5),1), Size(Dim(6),3), Size(Dim(7),3)),Float,NAnno)`, which doesn't have zero dim

File: src/test/scala/lms/transformation/test_distributed_tensor.scala [hunk 2]
```diff
@@ -228,5 +228,34 @@ class FixedSizeDistributedTensorTest extends TutorialFunSuite {
     checkWithLogPath("conv", driver.code, "cu", driver.setLogPath)
   }
 
+  test("dropout") {
```
Comment (luke-jiang) [inline src/test/scala/lms/transformation/test_distributed_tensor.scala:231]: the test case

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala [hunk 3]
```diff
@@ -111,6 +134,22 @@ trait FixedSizeDistributedTensorConvTypeLess extends FixedSizeDistributedTensorM
...
+        // save forward op in forwardNodes
+        forwardNodes += node
+        // save backward op in backwardNodes
+
+        (() => {
+          val x = new TENSOR(transform(a))
+          // val grads = gradMap.getGradsOfOp(s)
+          // val g = DropoutBackward(gradMap(s), grads(1), grads(2), params, anno, pos)
+          // Accumulate(gradMap(a), g, anno); ()
+          val grad = DropoutBackward(gradMap(s), gradMap(s), gradMap(s), params, anno, pos)
```
Comment (luke-jiang) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala:149]: dummy parameters to `DropoutBackward` for now

File: src/out/transformer/distributed_tensor/dropout.check.cu [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    size_t x40 = (size_t)0;
+    CUDNNCHECK(cudnnDropoutGetStatesSize(x8, &x40));
+    // end finding dropout backward states bytes
+    // begin creating dropout descriptor
+    cudnnDropoutDescriptor_t x41;
+    CUDNNCHECK(cudnnCreateDropoutDescriptor(&x41));
+    CUDNNCHECK(cudnnSetDropoutDescriptor(x41, x8, 0.5, x3, x40, 1));
+    // end creating dropout descriptor
+    // begin dropout backward pass
+    CUDNNCHECK(cudnnDropoutBackward(x8, x41, x37, x3, x37, x38, x3, x39));
```
Comment (luke-jiang) [inline src/out/transformer/distributed_tensor/dropout.check.cu:128]: should not be x3

File: src/out/transformer/distributed_tensor/dropout.check.cu [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  CUDA_CALL(cudaSetDevice(x7));
+  float* x19 = (float*)malloc(0 * sizeof(float));
+  CUDA_CALL(cudaMalloc(&x19, (size_t)(18 * sizeof(float))));
+  x13<<<dim3(28, 1, 1), dim3(512, 1, 1)>>>(x19, 0, 18);
+  // end initializing fixed GPU array of size 18 and type Float and device (pre-rename) x39
+  int x20 = 0;
+  while (x20 != 10) {
+    // begin initializing fixed GPU array of size 18 and type Float and device (pre-rename) x39
+    CUDA_CALL(cudaSetDevice(x7));
+    CUDA_CALL(cudaMalloc(&(float*)malloc(0 * sizeof(float)), (size_t)(18 * sizeof(float))));
```
Comment (luke-jiang) [inline src/out/transformer/distributed_tensor/dropout.check.cu:100]: this cudaMalloc is strange

File: src/out/transformer/distributed_tensor/dropout.check.cu [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  while (x19 != 10) {
+    // begin initializing fixed GPU array of size 18 and type Float and device (pre-rename) x39
+    CUDA_CALL(cudaSetDevice(x6));
+    float* x42 = (float*)malloc(0 * sizeof(float));
+    CUDA_CALL(cudaMalloc(&x42, (size_t)(18 * sizeof(float))));
+    x12<<<dim3(28, 1, 1), dim3(512, 1, 1)>>>(x42, 0, 18);
+    // end initializing fixed GPU array of size 18 and type Float and device (pre-rename) x39
+    // begin initializing fixed GPU array of size 18 and type Boolean and device (pre-rename) x39
+    CUDA_CALL(cudaSetDevice(x6));
+    bool* x43 = (bool*)malloc(0 * sizeof(bool));
```
Comment (luke-jiang) [inline src/out/transformer/distributed_tensor/dropout.check.cu:113]: x43 is d_states, allocated based on its dummy shape.

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala [hunk 7]
```diff
@@ -123,9 +156,23 @@ trait FixedSizeDistributedTensorOpsConv extends FixedSizeDistributedTensorOpsBas
...
+    def dropout(anno: Anno, params: DropoutParam = dropout_params_def)(implicit __pos: SourceContext): List[Rep[Tensor[T]]] = {
+      val op = DropoutForward(self, params, anno, __pos)
+      ((0 until 1): Range).toList.map(i => Wrap[Tensor[T]](TENSORS.getResult(op, i).x))
+    }
+
+    /*
+    def dropout(anno: Anno, params: DropoutParam = dropout_params_def)(implicit __pos: SourceContext): Rep[Tensor[T]] = {
+      val t = DropoutForward(self, params, anno, __pos)
+      Wrap[Tensor[T]](t.x)
+    }*/
```
Comment (feiwang3311) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala:176]: let's remove useless commented code

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCL.scala [hunk 8]
```diff
@@ -33,6 +33,9 @@ class DistributeTensor2MPI_NCCLAnalysis extends Traverser {
         case Node(s, op, _, _) if op.startsWith("tensor_conv") =>
             hasCudnn = true
             super.traverse(n)
+        case Node(s, op, _, _) if op.startsWith("tensors_dropout") =>
+            hasCudnn = true
+            super.traverse(n)
```
Comment (feiwang3311) [inline src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCL.scala:38]: maybe we should have a list of `cudnn_ops` and just do `if cudnn_ops.contains(op)`

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLBase.scala [hunk 9]
```diff
@@ -54,6 +54,12 @@ abstract class DistributeTensor2MPI_NCCLBase extends Transformer with MPIOps wit
     CUDA_SET_DEVICE(device)
     CUDA_MALLOC(size, m)
   }
+
+  // TODO: change it to better name, and add INT counterparts to other gpu functions as well
+  def gpu_array1_bytes(size: INT, m: Manifest[_], device: INT)(implicit __pos: SourceContext): ARRAY = {
```
Comment (feiwang3311) [inline src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLBase.scala:59]: would `gpu_array_by_byte` be a good name?

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala [hunk 10]
```diff
@@ -236,16 +236,121 @@ trait DistributeTensor2MPI_NCCLConv extends DistributeTensor2MPI_NCCLBase with C
...
+      // get input info and transform input tensors
+      val input_shape = tensor_shape(input, useOldMetadata = true)
+      val input_tensor = get_operand(input, anno)
+
+      val input_descriptor = getTensorDescriptor(input_shape, "tensor")
+      val output_descriptor = getTensorDescriptor(input_shape, "tensor")
+
+      generate_comment("begin finding dropout forward reserve space bytes") 
+      var reserve_bytes = 0
+      var reserve_bytes_v = VAR(SIZE_T(reserve_bytes))  // var read
```
Comment (feiwang3311) [inline src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala:266]: what if we just do `val reserve_bytes_v = VAR(SIZE_T(0))`?

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala [hunk 11]
```diff
@@ -236,16 +236,121 @@ trait DistributeTensor2MPI_NCCLConv extends DistributeTensor2MPI_NCCLBase with C
...
+      CUDNN_CHECK(CUDNN_DROPOUT_GET_RESERVE_SPACE_SZ(input_descriptor, reserve_bytes_v))
+      generate_comment("end finding dropout forward reserve space bytes") 
+
+      generate_comment("begin finding dropout forward states bytes")
+      var states_bytes = 0
+      var states_bytes_v = VAR(SIZE_T(reserve_bytes))  // var read
```
Comment (feiwang3311) [inline src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala:272]: similar here

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala (hunk 10 above)
Comment (luke-jiang) [reply src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala:266]: yeah reserve_bytes are not used.

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCL.scala (hunk 8 above)
Comment (luke-jiang) [reply src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCL.scala:38]: I was thinking the same thing. Maybe we can fix that after softmax and dropout are merged, together with other fixes.

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLBase.scala (hunk 9 above)
Comment (luke-jiang) [reply src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLBase.scala:59]: yeah, that's more descriptive.
btw I think we should also change the `array1` name. I think this fix can be made after the current branched are merged.

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala [hunk 12]
```diff
@@ -236,16 +236,118 @@ trait DistributeTensor2MPI_NCCLConv extends DistributeTensor2MPI_NCCLBase with C
...
+      val output = gpu_array(output_size, manifest[Float], myNCCLRank)
+      generate_comment("end allocating gpu array for the output of dropout")
+
+      generate_comment("begin dropout forward pass")
+      CUDNN_CHECK(CUDNN_DROPOUT_FWD(myCUDNNComm, dropout_descriptor, input_descriptor, new ARRAY(input_tensor), 
+        output_descriptor, output, d_reservespace, reserve_bytes_v))
+      generate_comment("end dropout forward pass")
+
+      // return dropout output
+      Adapter.g.reflect("tuple-view", output.x, d_states.x, d_reservespace.x)
```
[the remark this answers is not in the document]
Comment (luke-jiang) [reply src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala]: This is actually old version and has been fixed earlier. I'm not sure why it's not reflected 😂

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala (hunk 11 above)
Comment (luke-jiang) [reply src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala]: fixed in 9ea513c

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala (hunk 10 above)
Comment (luke-jiang) [reply src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala]: fixed in 9ea513c

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLBase.scala (hunk 9 above)
Comment (luke-jiang) [reply src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLBase.scala]: fixed in eee74d5

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala [hunk 13]
```diff
@@ -236,16 +236,118 @@ trait DistributeTensor2MPI_NCCLConv extends DistributeTensor2MPI_NCCLBase with C
...
+      val doutput_shape = tensor_shape(doutput, useOldMetadata = true)
+      val doutput_descriptor = getTensorDescriptor(doutput_shape, "tensor")
+      val dinput_descriptor = doutput_descriptor
+
+      val doutput_tensor = get_operand(doutput, anno)
+      val state_tensor = get_operand(state, anno)
+      val reserveSpace_tensor = get_operand(reserveSpace, anno)
+      
+      generate_comment("begin allocating gpu array for the gradient of input of dropout")
+      val dinput_size = doutput_shape(0) * doutput_shape(1) * doutput_shape(2) * doutput_shape(3)
```
[the remark this answers is not in the document]
Comment (luke-jiang) [reply src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala]: fixed in d2c2269

File: src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala [hunk 14]
```diff
@@ -236,16 +236,118 @@ trait DistributeTensor2MPI_NCCLConv extends DistributeTensor2MPI_NCCLBase with C
...
+      generate_comment("begin creating dropout descriptor")
+      val dropout_descriptor = new CUDNN_DROPOUT_DESCRIPTOR(NEW_STRUCT(manifest[CUDNN_DROPOUT_DESCRIPTOR], "cudnnDropoutDescriptor_t").x)
+      CUDNN_CHECK(CUDNN_CREATE_DROPOUT_DESCRIPTOR(dropout_descriptor))
+      CUDNN_CHECK(CUDNN_SET_DROPOUT_DESCRIPTOR(dropout_descriptor, myCUDNNComm, dropout, d_states, states_bytes_v, seed))
+      generate_comment("end creating dropout descriptor")
+
+      // allocate output tensor
+      // output tensor has the same shape as input tensor
+      generate_comment("begin allocating gpu array for the output of dropout")
+      val output_size = input_shape(0) * input_shape(1) * input_shape(2) * input_shape(3)
```
[the remark this answers is not in the document]
Comment (luke-jiang) [reply src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala]: fixed in d2c2269

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala (hunk 7 above)
Comment (luke-jiang) [reply src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala]: fixed in eee74d5

**Review by feiwang3311 — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**